File: pkg/common/backend.go
package common

import (
	"context"
	"io"
	"net"
	"strconv"
)

// Target describes the destination a backend connects to on behalf of a client.
type Target struct {
	// Network is the transport network ("tcp", "udp"). When empty, Net()
	// defaults to "tcp".
	Network string
	// Protocol is the detected application protocol. An empty value is treated
	// as ProtocolUnknown.
	Protocol Protocol
	// Host is the destination hostname or IP address literal.
	Host string
	// Port is the destination port.
	Port uint16
}

// Protocol identifies an application protocol observed by a frontend.
type Protocol string

const (
	// ProtocolAny is a capability wildcard and must not be used as a detected
	// Target protocol.
	ProtocolAny Protocol = "*"
	// ProtocolUnknown means the frontend could not identify the application
	// protocol.
	ProtocolUnknown Protocol = "unknown"
	// ProtocolHTTP identifies HTTP/1 or the HTTP/2 client connection preface.
	ProtocolHTTP Protocol = "http"
	// ProtocolTLS identifies a TLS client handshake.
	ProtocolTLS Protocol = "tls"
	// ProtocolDNS identifies DNS carried over TCP.
	ProtocolDNS Protocol = "dns"
)

// Capability describes one network/application-protocol combination accepted
// by a Backend. ProtocolAny matches every application protocol.
type Capability struct {
	Network  string
	Protocol Protocol
}

// SupportsNetwork reports whether capabilities contain the network.
func SupportsNetwork(capabilities []Capability, network string) bool {
	for _, capability := range capabilities {
		if capability.Network == network {
			return true
		}
	}
	return false
}

// SupportsAnyProtocol reports whether capabilities contain a wildcard entry
// for the network.
func SupportsAnyProtocol(capabilities []Capability, network string) bool {
	for _, capability := range capabilities {
		if capability.Network == network && capability.Protocol == ProtocolAny {
			return true
		}
	}
	return false
}

// Supports reports whether capabilities accept the target's network and
// application protocol.
func Supports(capabilities []Capability, target Target) bool {
	protocol := target.Protocol
	if protocol == "" {
		protocol = ProtocolUnknown
	}
	for _, capability := range capabilities {
		if capability.Network == target.Net() && (capability.Protocol == ProtocolAny || capability.Protocol == protocol) {
			return true
		}
	}
	return false
}

// Dialer establishes the transport connection used by a Backend. Frontends
// provide the dialer so they can control which network path backend traffic
// takes (for example, bypassing a TUN interface).
type Dialer interface {
	DialContext(ctx context.Context, network, address string) (net.Conn, error)
}

// DialFunc adapts a function to Dialer.
type DialFunc func(ctx context.Context, network, address string) (net.Conn, error)

// DialContext calls f.
func (f DialFunc) DialContext(ctx context.Context, network, address string) (net.Conn, error) {
	return f(ctx, network, address)
}

// SystemDialer returns a dialer that uses the host's normal routing table.
func SystemDialer() Dialer { return &net.Dialer{} }

// Address returns the "host:port" form suitable for net.Dial and HTTP CONNECT.
func (t Target) Address() string {
	return net.JoinHostPort(t.Host, strconv.Itoa(int(t.Port)))
}

// Net returns the network, defaulting to "tcp" when empty.
func (t Target) Net() string {
	if t.Network == "" {
		return "tcp"
	}
	return t.Network
}

// Backend dials a tunneled connection to a Target using the frontend-provided
// transport dialer. Implementations live in pkg/adapter/* (direct,
// httpproxy, socksproxy, ...). The returned io.ReadWriteCloser is handed to a
// ShimServer as its Backend.
type Backend interface {
	// Capabilities returns the network and application protocol combinations
	// that the backend can accept.
	Capabilities() []Capability
	Dial(ctx context.Context, target Target, dialer Dialer) (io.ReadWriteCloser, error)
}

File: rust/crates/puppy-core/tests/backend.rs
//! Tests for `backend.rs`.

use puppy_core::backend::{
	supports, supports_any_protocol, supports_network, Capability, Protocol, Target,
};

/// Verifies the capability-matching helpers (`supports_network`,
/// `supports_any_protocol`, `supports`) select backends per the documented
/// rules: exact network match, the `Any` (`*`) protocol wildcard, exact
/// protocol match, and normalization of an empty/`Unknown` target protocol
/// to `Unknown` (which never matches a non-wildcard capability).
#[test]
fn capabilities_match_spec_semantics() {
	let capabilities = vec![
		Capability {
			network: "tcp".to_string(),
			protocol: Protocol::Http,
		},
		Capability {
			network: "tcp".to_string(),
			protocol: Protocol::Dns,
		},
		Capability {
			network: "udp".to_string(),
			protocol: Protocol::Any,
		},
	];

	assert!(supports_network(&capabilities, "tcp"));
	assert!(!supports_network(&capabilities, "icmp"));

	// `SupportsAnyProtocol` returns true when a capability has the wildcard
	// `*` (Any) protocol — the udp entry qualifies, the tcp entries do not.
	assert!(!supports_any_protocol(&capabilities, "tcp"));
	assert!(supports_any_protocol(&capabilities, "udp"));

	assert!(supports(
		&capabilities,
		&Target {
			network: "tcp".to_string(),
			protocol: Protocol::Http,
			host: String::new(),
			port: 0,
		}
	));
	assert!(!supports(
		&capabilities,
		&Target {
			network: "tcp".to_string(),
			protocol: Protocol::Tls,
			host: String::new(),
			port: 0,
		}
	));
	assert!(supports(
		&capabilities,
		&Target {
			network: "tcp".to_string(),
			protocol: Protocol::Dns,
			host: String::new(),
			port: 0,
		}
	));
	// The udp wildcard accepts any application protocol marker.
	assert!(supports(
		&capabilities,
		&Target {
			network: "udp".to_string(),
			protocol: Protocol::Tls,
			host: String::new(),
			port: 0,
		}
	));
	// Empty protocol normalizes to Unknown, not Http, so no tcp capability
	// matches (only the udp wildcard would, on a different network).
	assert!(!supports(
		&capabilities,
		&Target {
			network: "tcp".to_string(),
			protocol: Protocol::Unknown,
			host: String::new(),
			port: 0,
		}
	));
}
